File: 백준/Platinum/2887. 행성 터널/행성 터널.py
# 특정 원소가 속한 집합을 찾기
def find_parent(parent, x):
    # 루트 노드를 찾을 때까지 재귀 호출
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

# 두 원소가 속한 집합을 찾기
def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

n = int(input()) # 행성의 개수
pos = [] # (x, y, z, 행성번호) 튜플형태로 위치저장
for i in range(1, n+1):
    x, y, z = map(int, input().split())
    pos.append((x, y, z, i))

# 간선을 저장할 리스트와 최종 비용을 담을 변수
edges = []
result = 0

# x, y, z좌표기준으로 정렬한 후, 인접한 행성끼리의 터널(간선)을 생성해 줌 (3 * (n-1)개의 간선 생성)
for axis in range(3):
    pos.sort(key=lambda item: (item[axis]))
    for i in range(n-1):
        cost = abs(pos[i+1][axis] - pos[i][axis])
        edges.append((cost, pos[i][3], pos[i+1][3]))

edges.sort()

# 부모 테이블 초기화
parent = [0] * (n+1)

# 부모 테이블상에서 부모를 자기 자신으로 초기화
for i in range(1, n+1):
    parent[i] = i

# 비용순으로 정렬된 간선을 하나씩 확인
for edge in edges:
    cost, a, b = edge

    # 사이클이 발생하지 않는 경우에만 집합에 포함
    if find_parent(parent, a) != find_parent(parent, b):
        union_parent(parent, a, b)
        result += cost
    
print(result)

File: 백준/Platinum/2887. 행성 터널/행성 터널.cs
class Program
{
    // 특정 원소가 속한 집합의 부모노드를 찾는 함수
    static int FindParent(int[] parentTable, int x)
    {
        if (parentTable[x] != x)
            parentTable[x] = FindParent(parentTable, parentTable[x]); // 부모노드의 번호를 경로압축방식으로 저장
        
        return parentTable[x];
    }

    // a, b가 속한 집합을 합치는 함수
    static void UnionParent(int[] parentTable, int a, int b)
    {
        a = FindParent(parentTable, a);
        b =  FindParent(parentTable, b);
        
        if (a < b)
            parentTable[b] = a;
        else
            parentTable[a] = b;
    }
    
    static void Main(string[] args)
    {
        // 입력
        int n = int.Parse(Console.ReadLine());
        
        (int x, int y, int z, int PlanetNumber)[] locations = new (int x, int y, int z, int PlanetNumber)[n];

        // 행성번호를 1~n번으로 하여 (x,y,z,행성번호) 형식으로 튜플배열에 저장
        for (int planetNumber = 0; planetNumber < n; planetNumber++)
        {
            var parts =  Console.ReadLine().Split(' ');
        
            int x = int.Parse(parts[0]);
            int y = int.Parse(parts[1]);
            int z = int.Parse(parts[2]);

            locations[planetNumber] = (x, y, z, planetNumber+1);
        }

        // 간선을 저장할 변수
        List<(int Cost, int PlanetA, int PlanetB)> edges = new List<(int Cost, int PlanetA, int PlanetB)>();

        // 모든 행성쌍에서 간선을 만들게되면 n*(n-1) / 2개의 간선이 만들어져서 메모리 초과가 나게 됨.
        // 그러므로 x, y, z위치를 기준으로 행성들을 정렬해서 인접한 행성들간에 간선을 생성(최소 신장 트리에 포함될 가능성이 있는 간선들)하여 메모리사용을 적게하여 해결
        // x, y, z좌표기준으로 정렬한 후, 인접한 행성끼리의 터널(간선)을 생성해 줌 (3 * (n-1)개의 간선 생성)
        var sortedLocationsX = locations.OrderBy(p => p.x).ToArray();
        for (int i=0; i<sortedLocationsX.Length - 1; i++)
        {
            var location = sortedLocationsX[i];
            var nextLocation = sortedLocationsX[i + 1];
            var cost = Math.Abs(nextLocation.x - location.x);
            edges.Add((cost, location.PlanetNumber, nextLocation.PlanetNumber));
        }
        
        var sortedLocationsY = locations.OrderBy(p => p.y).ToArray();
        for (int i=0; i<sortedLocationsY.Length - 1; i++)
        {
            var location = sortedLocationsY[i];
            var nextLocation = sortedLocationsY[i + 1];
            var cost = Math.Abs(nextLocation.y - location.y);
            edges.Add((cost, location.PlanetNumber, nextLocation.PlanetNumber));
        }
        
        var sortedLocationsZ = locations.OrderBy(p => p.z).ToArray();
        for (int i=0; i<sortedLocationsZ.Length - 1; i++)
        {
            var location = sortedLocationsZ[i];
            var nextLocation = sortedLocationsZ[i + 1];
            var cost = Math.Abs(nextLocation.z - location.z);
            edges.Add((cost, location.PlanetNumber, nextLocation.PlanetNumber));
        }
        
        var sortedEdges = edges.OrderBy(edge => edge.Cost).ToList();
        
        // 부모테이블 초기화
        int[] parentTable = new int[n + 1];

        for (int i = 1; i < n+1; i++)
        {
            parentTable[i] = i;
        }
        
        int result = 0;
        // 크루스칼 알고리즘으로 최소비용트리 생성
        foreach (var edge in sortedEdges)
        {
            var cost =  edge.Cost;
            var a =  edge.PlanetA;
            var b =  edge.PlanetB;
            
            // 사이클이 발생하지 않는경우에만 집합에 포함
            if (FindParent(parentTable, a) != FindParent(parentTable, b))
            {
                UnionParent(parentTable, a, b);
                result += cost;
            }
        }
        
        Console.WriteLine(result);
    }
}
